src/service.ts

```diff
@@ -55,175 +62,177 @@ interface ServiceOptions {
+  private projectId: string;
+  private projectIdRequired: boolean;
+  private Promise: Promise<{}>;
+  // TODO: make this private
+  makeAuthenticatedRequest;
+  // TODO: make this private
+  authClient;
+  private getCredentials;
+
+  /**
+   * Service is a base class, meant to be inherited from by a "service," like
+   * BigQuery or Storage.
+   *
+   * This handles making authenticated requests by exposing a `makeReq_` function.
+   *
+   * @constructor
+   * @alias module:common/service
+   *
+   * @param {object} config - Configuration object.
+   * @param {string} config.baseUrl - The base URL to make API requests to.
+   * @param {string[]} config.scopes - The scopes required for the request.
+   * @param {object=} options - [Configuration object](#/docs).
+   */
+  constructor(config: ServiceConfig, options?: ServiceOptions) {
+    options = options || {};
+
+    this.baseUrl = config.baseUrl;
+    this.globalInterceptors = arrify(options.interceptors_);
+    this.interceptors = [];
+    this.packageJson = config.packageJson;
+    this.projectId = options.projectId || PROJECT_ID_TOKEN;
+    this.projectIdRequired = config.projectIdRequired !== false;
+    this.Promise = options.promise || Promise;
+
+    const reqCfg = extend({}, config, {
+      projectIdRequired: this.projectIdRequired,
+      projectId: this.projectId,
+      credentials: options.credentials,
+      keyFile: options.keyFilename,
+      email: options.email,
+      token: options.token,
     });
-  }
-}
 
-/**
- * Get and update the Service's project ID.
- *
- * @param {function} callback - The callback function.
- */
-Service.prototype.getProjectId = function(callback: (err: Error|null, projectId?: string) => void) {
-  const self = this;
+    this.makeAuthenticatedRequest = util.makeAuthenticatedRequestFactory(reqCfg);
+    this.authClient = this.makeAuthenticatedRequest.authClient;
+    this.getCredentials = this.makeAuthenticatedRequest.getCredentials;
+
+    const isCloudFunctionEnv = !!process.env.FUNCTION_NAME;
 
-  this.authClient.getProjectId((err: Error|null, projectId: string) => {
-    if (err) {
-      callback(err);
-      return;
+    if (isCloudFunctionEnv) {
+      this.interceptors.push({
+        request(reqOpts: r.Options) {
+          reqOpts.forever = false;
+          return reqOpts;
+        },
+      });
     }
+  }
 
-    if (self.projectId === PROJECT_ID_TOKEN && projectId) {
-      self.projectId = projectId;
+  /**
+   * Get and update the Service's project ID.
+   *
+   * @param {function} callback - The callback function.
+   */
+  getProjectId(): Promise<string>;
+  getProjectId(callback: (err: Error|null, projectId?: string) => void): void;
+  getProjectId(callback?: (err: Error|null, projectId?: string) => void): Promise<string>|void {
+    if (!callback) {
+      return this.getProjectIdAsync();
     }
+    this.getProjectIdAsync()
+        .then(p => callback(null, p))
+        .catch(e => callback(e));
```

Comment:
If `callback` throws in `then` it will be called a second time in the `catch` handler. We should use `.then(p => callback(null, p), e => callback(e))`

Reply:
Woah, TIL.  If I use the rejected handler in here, should I also have a `.catch`?

Reply:
I don't think there needs to be a `.catch` since it's user code that's would be throwing.